README.md
## Arc Progress Bar

The Arc Progress Bar is an open source Flutter package that simplifies the creation of a curved or arc-shaped progress
bar widget. It offers a wide range of customization options for users, taking care of the intricate curve calculations
and animations on your behalf.

## Add Dependency

Add this to your pubspec.yaml dependencies:

```yaml
dependencies:
  arc_progress_bar: ^1.0.4
```

## How to use

First of all make sure to call the import

```dart
import 'package:arc_progress_bar_new/arc_progress_bar_new.dart';
```


Then just customize the widget's parameters.

```dart
  ArcProgressBar(
    percentage: _progressPercentage,
    arcThickness: 5,
    innerPadding: 16,
    animateFromLastPercent: true,
    handleSize: 10,
    backgroundColor: Colors.black12,
    foregroundColor: Colors.black
    )
```

<img src="https://raw.githubusercontent.com/Frankline-Sable/arc_progress_bar/main/example/screenshots/1.png" alt="Colored Nodes & Outlines" width="500"/>

---

## Screenshots & Implementations

### Example 1 - With custom Colors and Handle

<img src="https://raw.githubusercontent.com/Frankline-Sable/arc_progress_bar/main/example/screenshots/2.png" alt="With custom colors and handle" width="300"/>

```dart
 ArcProgressBar(
    percentage: _progressPercentage,
    arcThickness: 15,
    innerPadding: 48,
    strokeCap: StrokeCap.round,
    handleSize: 50,
    handleWidget: Container(
      decoration: const BoxDecoration(color: Colors.red)),
    foregroundColor: Colors.redAccent,
    backgroundColor: Colors.red.shade100,   
 ),
```

<br>

### Example 2 - With Icon And Texts

<img src="https://raw.githubusercontent.com/Frankline-Sable/arc_progress_bar/main/example/screenshots/3.png" alt="With Icons And Texts" width="300"/>

```dart
 ArcProgressBar(
    percentage: _progressPercentage,
    bottomLeftWidget: const Text("Level 3"),
    bottomRightWidget: const Text("240/300"),
    bottomCenterWidget: const Text("RECRUIT"),
    centerWidget: Image.asset("assets/images/insignia.png",
        height: 200, width: 200, fit: BoxFit.contain)),   
 ),
```

<br>

### Example 3 - Very Simple & Customizable

<img src="https://raw.githubusercontent.com/Frankline-Sable/arc_progress_bar/main/example/screenshots/4.png" alt="Very Customizable" width="300"/>

```dart
 ArcProgressBar(
   percentage: _progressPercentage,
   arcThickness: 5,
   innerPadding: 16,
   animateFromLastPercent: true,
   handleSize: 10,
   backgroundColor: Colors.black12,
   foregroundColor: Colors.black),
```

<br>


Very customizable, feel free to customize however you like! 😎


CHANGELOG.md
## 0.0.1 - 11 October 2023

* Initial release
* Additional features coming soon

## 1.0.0 - 11 October 2023

* Updated release

## 1.0.1 - 11 October 2023

* Updated docs

## 1.0.2 - 11 October 2023

* Updated docs
* Did some housekeeping

## 1.0.3 - 11 October 2023

* Updated 

## 1.0.4 - 11 October 2023

* Updated the doc

test/arc_progress_bar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:arc_progress_bar/arc_progress_bar.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/src/arc_progress_bar.dart
import 'package:arc_progress_bar/src/classes/arc_paint.dart';
import 'package:flutter/material.dart';
import 'dart:math' as math;
import 'package:vector_math/vector_math.dart' as vector;
import 'extensions/nums.dart';

/// The arc progress bar widget
class ArcProgressBar extends StatefulWidget {
  const ArcProgressBar(
      {Key? key,
      this.percentage = 0,
      this.innerPadding = 16,
      this.animateFromLastPercent = true,
      this.animationDuration = const Duration(seconds: 1),
      this.animationCurve = Curves.easeInOutCubic,
      this.handleSize,
      this.handleWidget,
      this.handleColor,
      this.arcThickness = 10,
      this.strokeCap = StrokeCap.square,
      this.backgroundColor = Colors.black12,
      this.foregroundColor = Colors.black,
      this.bottomCenterWidget,
      this.bottomRightWidget,
      this.bottomLeftWidget,
      this.centerWidget,
      this.onAnimationStart,
      this.onAnimationEnd})
      : super(key: key);

  /// The current progress of the arc in percentage
  final double percentage;

  /// The size of the handle
  final double? handleSize;

  /// Widget to replace the handle
  final Widget? handleWidget;

  /// The color of the handle
  final Color? handleColor;

  /// Internal padding of the widget
  final double innerPadding;

  /// Arc line thickness
  final double arcThickness;

  /// StrokeCap for the curve ends
  final StrokeCap strokeCap;

  /// Background color of the curve
  final Color backgroundColor;

  /// Foreground color of the curve
  final Color foregroundColor;

  /// Widget to display at the center of the curve
  final Widget? centerWidget;

  /// Widget to display at the bottom left of the curve
  final Widget? bottomLeftWidget;

  /// Widget to display at the bottom right of the curve
  final Widget? bottomRightWidget;

  /// Widget to display at the bottom center of the curve
  final Widget? bottomCenterWidget;

  /// Action after the progress change animation
  final Function? onAnimationEnd;

  /// Action before the progress change animation
  final Function? onAnimationStart;

  // Curve for the progress change Animation
  final Curve animationCurve;

  // Animation duration for the progress change
  final Duration animationDuration;

  // Preserve the previous state of the animation
  final bool animateFromLastPercent;

  @override
  State<ArcProgressBar> createState() => _ArcProgressBarState();
}

class _ArcProgressBarState extends State<ArcProgressBar>
    with SingleTickerProviderStateMixin {
  late Animation<double> animation;
  late AnimationController _animationController;

  double _handleSize = 10;
  double _percent = 0.0;
  double canvasSize = 300;
  Color _handleColor = Colors.black;

//
  @override
  Widget build(BuildContext context) {
    _handleSize = widget.handleSize ?? (widget.arcThickness * 2);
    _handleColor = widget.handleColor ?? widget.foregroundColor;
    // bug check!!
    _percent = _percent.isNaN ? 0 : _percent;

    // Converting the percentage progress to degrees, NB: since the arc is 180
    final sweep = _percent.clamp(0, 100) * (180 / 100);
    return LayoutBuilder(builder: (ctx, constraints) {
      double radiusForHandle = (constraints.maxWidth) / 2;
      radiusForHandle = radiusForHandle - (_handleSize / 2);

      if (constraints.maxWidth < 500) canvasSize = constraints.maxWidth;
      return Stack(
        children: [
          CustomPaint(
            size: Size(
              canvasSize,
              canvasSize / 1.8,
            ),
            painter: ArcPaint(vector.radians((180)), widget.backgroundColor,
                widget.arcThickness, widget.strokeCap,
                padding: widget.innerPadding),
          ),
          CustomPaint(
            size: Size(
              canvasSize,
              canvasSize / 1.65,
            ),
            painter: ArcPaint(vector.radians((sweep)), widget.foregroundColor,
                widget.arcThickness, widget.strokeCap,
                padding: widget.innerPadding),
          ),
          if (widget.bottomLeftWidget != null)
            Positioned(
                bottom: widget.innerPadding,
                left: 0,
                child: widget.bottomLeftWidget!),
          if (widget.bottomRightWidget != null)
            Positioned(
                bottom: widget.innerPadding,
                right: 0,
                child: widget.bottomRightWidget!),
          if (widget.bottomCenterWidget != null)
            Positioned(
                bottom: widget.innerPadding,
                left: 0,
                right: 0,
                child: Align(
                  alignment: AlignmentDirectional.center,
                  child: widget.bottomCenterWidget!,
                )),
          if (widget.centerWidget != null)
            Positioned(
                right: 0,
                left: 0,
                top: 0,
                bottom: 0,
                child: Align(
                    alignment: AlignmentDirectional.center,
                    child: widget.centerWidget!)),
          Positioned(
              left: ((radiusForHandle) -
                  ((((radiusForHandle +
                          (_handleSize / 2) -
                          widget.innerPadding) *
                      math.sin(vector.radians(
                          (_percent.clampRange(min: 90, max: -90)))))))),
              top: ((radiusForHandle) -
                  ((((radiusForHandle +
                          (_handleSize / 2) -
                          widget.innerPadding) *
                      math.cos(vector.radians(
                          (_percent.clampRange(min: 90, max: -90)))))))),
              child: Container(
                width: _handleSize,
                height: _handleSize,
                decoration: widget.handleWidget == null
                    ? BoxDecoration(
                        color: _handleColor,
                        shape: BoxShape.circle,
                      )
                    : null,
                child: widget.handleWidget,
              )),
        ],
      );
    });
  }

  @override
  void initState() {
    _animationController =
        AnimationController(duration: widget.animationDuration, vsync: this);
    final curvedAnimation = CurvedAnimation(
        parent: _animationController, curve: widget.animationCurve);
    animation =
        Tween<double>(begin: 0, end: widget.percentage).animate(curvedAnimation)
          ..addListener(() {
            switch (_animationController.status) {
              case AnimationStatus.forward:
                if (widget.onAnimationStart != null) {
                  widget.onAnimationStart!();
                }
                break;
              case AnimationStatus.completed:
                if (widget.onAnimationEnd != null) {
                  widget.onAnimationEnd!();
                }
                break;
              default:
            }
            setState(() {
              _percent = animation.value;
            });
          });

    _animationController.forward();
    super.initState();
  }

  @override
  dispose() {
    _animationController.dispose(); // you need this
    super.dispose();
  }

  @override
  void didUpdateWidget(ArcProgressBar oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.percentage != widget.percentage) {
      _animationController.duration = widget.animationDuration;
      animation = Tween(
              begin: widget.animateFromLastPercent ? oldWidget.percentage : 0.0,
              end: widget.percentage)
          .animate(
        CurvedAnimation(
            parent: _animationController, curve: Curves.easeInOutCubic),
      );
      _animationController.forward(from: 0.0);
    }
  }
}


lib/src/classes/arc_paint.dart
import 'package:flutter/material.dart';
import 'package:vector_math/vector_math.dart' as vector;
import 'dart:math' as math;

class ArcPaint extends CustomPainter {
  double arc;
  double strokeWidth;
  Color curveColor;
  double padding;
  StrokeCap strokeCap;

  ArcPaint(
    this.arc,
    this.curveColor,
    this.strokeWidth,
    this.strokeCap, {
    this.padding = 0,
  });

  @override
  void paint(Canvas canvas, Size size) {
    final rect = Rect.fromLTRB(
        padding, padding, size.width - padding, size.width - padding);
    final startAngle = -math.pi + vector.radians(0);

    final sweepAngle = arc;
    const useCenter = false;
    final paint = Paint()
      ..strokeCap = strokeCap
      ..color = curveColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth;

    canvas.drawArc(rect, startAngle, sweepAngle, useCenter, paint);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}


lib/src/arc_progress_bar_new.dart
import 'package:flutter/material.dart';
import 'dart:math' as math;
import 'package:vector_math/vector_math.dart' as vector;
import 'classes/arc_paint.dart';
import 'extensions/nums.dart';

/// The arc progress bar widget
class ArcProgressBar extends StatefulWidget {
  const ArcProgressBar(
      {Key? key,
      this.percentage = 0,
      this.innerPadding = 16,
      this.animateFromLastPercent = true,
      this.animationDuration = const Duration(seconds: 1),
      this.animationCurve = Curves.easeInOutCubic,
      this.handleSize,
      this.handleWidget,
      this.handleColor,
      this.arcThickness = 10,
      this.strokeCap = StrokeCap.square,
      this.backgroundColor = Colors.black12,
      this.foregroundColor = Colors.black,
      this.bottomCenterWidget,
      this.bottomRightWidget,
      this.bottomLeftWidget,
      this.centerWidget,
      this.onAnimationStart,
      this.onAnimationEnd})
      : super(key: key);

  /// The current progress of the arc in percentage
  final double percentage;

  /// The size of the handle
  final double? handleSize;

  /// Widget to replace the handle
  final Widget? handleWidget;

  /// The color of the handle
  final Color? handleColor;

  /// Internal padding of the widget
  final double innerPadding;

  /// Arc line thickness
  final double arcThickness;

  /// StrokeCap for the curve ends
  final StrokeCap strokeCap;

  /// Background color of the curve
  final Color backgroundColor;

  /// Foreground color of the curve
  final Color foregroundColor;

  /// Widget to display at the center of the curve
  final Widget? centerWidget;

  /// Widget to display at the bottom left of the curve
  final Widget? bottomLeftWidget;

  /// Widget to display at the bottom right of the curve
  final Widget? bottomRightWidget;

  /// Widget to display at the bottom center of the curve
  final Widget? bottomCenterWidget;

  /// Action after the progress change animation
  final Function? onAnimationEnd;

  /// Action before the progress change animation
  final Function? onAnimationStart;

  // Curve for the progress change Animation
  final Curve animationCurve;

  // Animation duration for the progress change
  final Duration animationDuration;

  // Preserve the previous state of the animation
  final bool animateFromLastPercent;

  @override
  State<ArcProgressBar> createState() => _ArcProgressBarState();
}

class _ArcProgressBarState extends State<ArcProgressBar>
    with SingleTickerProviderStateMixin {
  late Animation<double> animation;
  late AnimationController _animationController;

  double _handleSize = 10;
  double _percent = 0.0;
  double canvasSize = 300;
  Color _handleColor = Colors.black;

//
  @override
  Widget build(BuildContext context) {
    _handleSize = widget.handleSize ?? (widget.arcThickness * 2);
    _handleColor = widget.handleColor ?? widget.foregroundColor;
    // bug check!!
    _percent = _percent.isNaN ? 0 : _percent;

    // Converting the percentage progress to degrees, NB: since the arc is 180
    final sweep = _percent.clamp(0, 100) * (180 / 100);
    return LayoutBuilder(builder: (ctx, constraints) {
      double radiusForHandle = (constraints.maxWidth) / 2;
      radiusForHandle = radiusForHandle - (_handleSize / 2);

      if (constraints.maxWidth < 500) canvasSize = constraints.maxWidth;
      return Stack(
        children: [
          CustomPaint(
            size: Size(
              canvasSize,
              canvasSize / 1.8,
            ),
            painter: ArcPaint(vector.radians((180)), widget.backgroundColor,
                widget.arcThickness, widget.strokeCap,
                padding: widget.innerPadding),
          ),
          CustomPaint(
            size: Size(
              canvasSize,
              canvasSize / 1.65,
            ),
            painter: ArcPaint(vector.radians((sweep)), widget.foregroundColor,
                widget.arcThickness, widget.strokeCap,
                padding: widget.innerPadding),
          ),
          if (widget.bottomLeftWidget != null)
            Positioned(
                bottom: widget.innerPadding,
                left: 0,
                child: widget.bottomLeftWidget!),
          if (widget.bottomRightWidget != null)
            Positioned(
                bottom: widget.innerPadding,
                right: 0,
                child: widget.bottomRightWidget!),
          if (widget.bottomCenterWidget != null)
            Positioned(
                bottom: widget.innerPadding,
                left: 0,
                right: 0,
                child: Align(
                  alignment: AlignmentDirectional.center,
                  child: widget.bottomCenterWidget!,
                )),
          if (widget.centerWidget != null)
            Positioned(
                right: 0,
                left: 0,
                top: 0,
                bottom: 0,
                child: Align(
                    alignment: AlignmentDirectional.center,
                    child: widget.centerWidget!)),
          Positioned(
              left: ((radiusForHandle) -
                  ((((radiusForHandle +
                          (_handleSize / 2) -
                          widget.innerPadding) *
                      math.sin(vector.radians(
                          (_percent.clampRange(min: 90, max: -90)))))))),
              top: ((radiusForHandle) -
                  ((((radiusForHandle +
                          (_handleSize / 2) -
                          widget.innerPadding) *
                      math.cos(vector.radians(
                          (_percent.clampRange(min: 90, max: -90)))))))),
              child: Container(
                width: _handleSize,
                height: _handleSize,
                decoration: widget.handleWidget == null
                    ? BoxDecoration(
                        color: _handleColor,
                        shape: BoxShape.circle,
                      )
                    : null,
                child: widget.handleWidget,
              )),
        ],
      );
    });
  }

  @override
  void initState() {
    _animationController =
        AnimationController(duration: widget.animationDuration, vsync: this);
    final curvedAnimation = CurvedAnimation(
        parent: _animationController, curve: widget.animationCurve);
    animation =
        Tween<double>(begin: 0, end: widget.percentage).animate(curvedAnimation)
          ..addListener(() {
            switch (_animationController.status) {
              case AnimationStatus.forward:
                if (widget.onAnimationStart != null) {
                  widget.onAnimationStart!();
                }
                break;
              case AnimationStatus.completed:
                if (widget.onAnimationEnd != null) {
                  widget.onAnimationEnd!();
                }
                break;
              default:
            }
            setState(() {
              _percent = animation.value;
            });
          });

    _animationController.forward();
    super.initState();
  }

  @override
  dispose() {
    _animationController.dispose(); // you need this
    super.dispose();
  }

  @override
  void didUpdateWidget(ArcProgressBar oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.percentage != widget.percentage) {
      _animationController.duration = widget.animationDuration;
      animation = Tween(
              begin: widget.animateFromLastPercent ? oldWidget.percentage : 0.0,
              end: widget.percentage)
          .animate(
        CurvedAnimation(
            parent: _animationController, curve: Curves.easeInOutCubic),
      );
      _animationController.forward(from: 0.0);
    }
  }
}


example/lib/main.dart
import 'package:arc_progress_bar_new/arc_progress_bar_new.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: ArcProgressBarApp(),
    );
  }
}

class ArcProgressBarApp extends StatefulWidget {
  const ArcProgressBarApp({super.key});

  @override
  State<ArcProgressBarApp> createState() => _ArcProgressBarAppState();
}

class _ArcProgressBarAppState extends State<ArcProgressBarApp> {
  double _progressPercentage = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Arc Progress Bar Demo"),
      ),
      body: SingleChildScrollView(
        child: Column(
          children: <Widget>[
            // With icons and texts widgets
            ArcProgressBar(
                percentage: _progressPercentage,
                bottomLeftWidget: const Text("Level 3"),
                bottomRightWidget: const Text("240/300"),
                bottomCenterWidget: const Text("RECRUIT"),
                centerWidget: Image.asset("assets/images/insignia.png",
                    height: 200, width: 200, fit: BoxFit.contain)),

            // With different colors
            ArcProgressBar(
              percentage: _progressPercentage,
              arcThickness: 15,
              innerPadding: 48,
              strokeCap: StrokeCap.round,
              handleSize: 50,
              handleWidget: Container(
                decoration: const BoxDecoration(color: Colors.red),
              ),
              foregroundColor: Colors.redAccent,
              backgroundColor: Colors.red.shade100,
            ),

            // Simple arc progress bar
            ArcProgressBar(
              percentage: _progressPercentage,
              arcThickness: 5,
              foregroundColor: Colors.blue,
              innerPadding: 64,
            ),

            const Padding(
              padding: EdgeInsets.symmetric(vertical: 32),
              child: Divider(
                color: Colors.black38,
              ),
            ),
            Slider(
                value: _progressPercentage,
                min: 0,
                max: 100,
                onChanged: (val) {
                  setState(() {
                    _progressPercentage = val;
                  });
                }),
          ],
        ),
      ),
    );
  }
}
